File: src/sync.js
#!/usr/bin/env node

const axios = require("axios");
const { Temporal } = require("temporal-polyfill");
const { Command } = require("commander");
const assert = require("node:assert");
const path = require("path");

require("dotenv").config({ path: path.resolve(__dirname, "../.env") });

// --- CONFIGURATION ---
const {
  CLOCKIFY_API_KEY,
  YOUTRACK_API_KEY,
  CLOCKIFY_WORKSPACE_ID,
  CLOCKIFY_USER_ID,
} = process.env;

const CLOCKIFY_URL = `https://api.clockify.me/api/v1/workspaces/${CLOCKIFY_WORKSPACE_ID}/user/${CLOCKIFY_USER_ID}/time-entries`;
const YOUTRACK_BASE_URL = "https://issues.yournextagency.com/api";

// --- API CALLS ---
const youTrackApi = axios.create({
  baseURL: YOUTRACK_BASE_URL,
  headers: {
    Authorization: `Bearer ${YOUTRACK_API_KEY}`,
  },
});

const clockifyApi = axios.create({
  baseURL: CLOCKIFY_URL,
  headers: {
    "X-Api-Key": CLOCKIFY_API_KEY,
  },
});

async function getBillableEntries(dateRange = {}) {
  const { data } = await clockifyApi.get("", {
    headers: { "X-Api-Key": CLOCKIFY_API_KEY },
    params: { ...dateRange, hydrated: true },
  });

  return (data || []).filter((entry) => entry.billable);
}

async function getIssueWorkItems(issueId) {
  const items = [];
  let skip = 0;
  const limit = 5000;

  while (true) {
    const { data } = await youTrackApi.get(
      `/issues/${issueId}/timeTracking/workItems?fields=text,date&$skip=${skip}&$top=${limit}`,
    );

    items.push(...data);
    if (data.length < limit) break;
    skip += limit;
  }

  return items;
}

async function postIssueWorkItem(
  issueId,
  { durationMinutes, workTypeId, description, date },
) {
  assert.match(description, /\[[a-z\d]{24}]/);

  const payload = {
    duration: { minutes: durationMinutes },
    text: description,
    date: date,
  };

  if (workTypeId) {
    payload.type = { id: workTypeId };
  }

  const { data } = await youTrackApi.post(
    `/issues/${issueId}/timeTracking/workItems`,
    payload,
  );

  return data;
}

async function getAllWorkItemTypes() {
  const { data: workTypes } = await youTrackApi.get(
    `/admin/timeTrackingSettings/workItemTypes?fields=id,name`,
  );

  return workTypes.reduce(
    (acc, type) => ({
      ...acc,
      [type.name.toLowerCase().trim()]: type.id,
    }),
    {},
  );
}

// --- HELPERS ---
function extractIssueId(text) {
  const match = text.trim().match(/^([A-Z0-9]+-\d+)/);

  return match ? match[1] : null;
}

function extractInternalId(text) {
  const match = text.trim().match(/\[([a-z\d]{24})]/);

  return match ? match[1] : null;
}

function createTimeEntry(billableEntry) {
  const { id, description, tags, task, timeInterval } = billableEntry;

  return {
    id: id,
    description: description,
    type: tags?.[0]?.name.toLowerCase().trim() || null,
    issueId: task?.name ? extractIssueId(task.name) : null,
    startDate: timeInterval.start ? new Date(timeInterval.start) : null,
    durationMinutes: timeInterval.duration
      ? Temporal.Duration.from(timeInterval.duration).total("minutes")
      : null,
  };
}

function getUniqueBy(arr, keyFn) {
  return Array.from(new Set(arr.map(keyFn).filter(Boolean)));
}

function getOptionDateRange(now, options) {
  const today = now.startOfDay();

  if (options.today) {
    return {
      start: today,
      end: today.add({ days: 1 }),
    };
  }

  if (options.lastNDays) {
    const start = today.subtract({ days: options.lastNDays });
    return {
      start: start,
      end: today,
    };
  }

  // Yesterday is the default
  return {
    start: today.subtract({ days: 1 }),
    end: today,
  };
}

// --- MAIN ---
async function sync() {
  const program = new Command();

  program
    .option("--today", "Sync only today's entries")
    .option("--yesterday", "Sync only yesterday's entries")
    .option(
      "--last-n-days <n>",
      "Sync entries from the last N days (excluding today)",
      parseInt,
    )
    .option("--dry-run", "Print what would be synced without posting anything")
    .parse();

  const options = program.opts();

  if (options.dryRun) {
    console.log("⚙️ Running in dry-run mode");
  }

  const now = Temporal.Now.zonedDateTimeISO("UTC");
  const { start, end } = getOptionDateRange(now, options);
  const dateRange = {
    start: start.toInstant().toString(),
    end: end.toInstant().toString(),
  };

  console.log(
    `⏳ Syncing entries from ${start.toLocaleString()} to ${end.toLocaleString()}`,
  );

  const entries = await getBillableEntries(dateRange);
  const workTypes = await getAllWorkItemTypes();

  const timeEntries = entries.map(createTimeEntry);

  const uniqueIssueIds = getUniqueBy(
    timeEntries.filter((entry) => entry.issueId),
    (entry) => entry.issueId,
  );

  const workItemsByIssueId = new Map(
    await Promise.all(
      uniqueIssueIds.map(async (issueId) => {
        const issueWorkItems = await getIssueWorkItems(issueId);
        const workItemInternalIds = issueWorkItems
          .map((item) => extractInternalId(item.text || ""))
          .filter(Boolean);

        return [issueId, new Set(workItemInternalIds)];
      }),
    ),
  );

  for (const entry of timeEntries) {
    const displayDate = entry.startDate.toDateString();

    if (!entry.issueId) {
      console.warn(`⚠️ No Issue: ${entry.description} on ${displayDate}`);
      continue;
    }

    const existingItems = workItemsByIssueId.get(entry.issueId);
    const exists = existingItems.has(entry.id);

    if (exists) {
      console.log(
        `⏩ Skipping: ${entry.description} (${entry.issueId}) on ${displayDate}`,
      );
      continue;
    }

    const data = {
      description: `${entry.description} [${entry.id}]`,
      date: entry.startDate.getTime(),
      durationMinutes: entry.durationMinutes,
    };

    if (entry.type && workTypes[entry.type]) {
      data.workTypeId = workTypes[entry.type];
    }

    if (!options.dryRun) {
      await postIssueWorkItem(entry.issueId, data);
    }
    console.log(
      `✅ Synced: ${entry.description} (${entry.issueId}) on ${displayDate}`,
    );
  }
}

sync().catch((err) => console.error("🛑️ Sync failed:", err));

File: sync.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
	"time"
)

type timeEntry struct {
	id              string
	description     string
	entryType       string
	issueID         string
	startDate       time.Time
	durationMinutes int
}

var (
	issueIDRe     = regexp.MustCompile(`^([A-Z0-9]+-\d+)`)
	timeEntryIDRe = regexp.MustCompile(`\[([a-z\d]{24})\]`)
)

func extractIssueID(text string) string {
	m := issueIDRe.FindStringSubmatch(strings.TrimSpace(text))
	if m == nil {
		return ""
	}
	return m[1]
}

func extractTimeEntryID(text string) string {
	m := timeEntryIDRe.FindStringSubmatch(strings.TrimSpace(text))
	if m == nil {
		return ""
	}
	return m[1]
}

func toTimeEntry(e clockifyEntry) timeEntry {
	te := timeEntry{
		id:          e.ID,
		description: e.Description,
	}
	if len(e.Tags) > 0 {
		te.entryType = strings.ToLower(strings.TrimSpace(e.Tags[0].Name))
	}
	if e.Task != nil {
		te.issueID = extractIssueID(e.Task.Name)
	}
	if e.TimeInterval.Start != "" {
		if t, err := time.Parse(time.RFC3339, e.TimeInterval.Start); err == nil {
			te.startDate = t
		}
	}
	if e.TimeInterval.Duration != "" {
		if m, err := parseISO8601DurationMinutes(e.TimeInterval.Duration); err == nil {
			te.durationMinutes = m
		}
	}
	return te
}

type dateRange struct {
	start time.Time
	end   time.Time
}

func getOptionDateRange(now time.Time, opts syncOptions) dateRange {
	today := now.Truncate(24 * time.Hour)

	switch {
	case opts.today:
		return dateRange{start: today, end: today.AddDate(0, 0, 1)}
	case opts.lastNDays > 0:
		return dateRange{start: today.AddDate(0, 0, -opts.lastNDays), end: today}
	default:
		return dateRange{start: today.AddDate(0, 0, -1), end: today}
	}
}

func syncEntries(clockify *clockifyClient, youtrack *youtrackClient, now time.Time, opts syncOptions) error {
	rng := getOptionDateRange(now, opts)

	fmt.Printf("⏳ Syncing entries from %s to %s\n",
		rng.start.Local().Format("Mon Jan 02 2006 15:04:05 MST"),
		rng.end.Local().Format("Mon Jan 02 2006 15:04:05 MST"),
	)

	entries, err := clockify.getBillableEntries(
		rng.start.UTC().Format(time.RFC3339),
		rng.end.UTC().Format(time.RFC3339),
	)
	if err != nil {
		return fmt.Errorf("fetch billable entries: %w", err)
	}

	workTypes, err := youtrack.getAllWorkItemTypes()
	if err != nil {
		return fmt.Errorf("fetch work item types: %w", err)
	}

	timeEntries := make([]timeEntry, 0, len(entries))
	for _, e := range entries {
		timeEntries = append(timeEntries, toTimeEntry(e))
	}

	seenIssues := make(map[string]bool)
	for _, te := range timeEntries {
		if te.issueID != "" {
			seenIssues[te.issueID] = true
		}
	}

	workItemsByIssue := make(map[string]map[string]bool, len(seenIssues))
	for issueID := range seenIssues {
		items, err := youtrack.getIssueWorkItems(issueID)
		if err != nil {
			return fmt.Errorf("fetch work items for %s: %w", issueID, err)
		}
		idSet := make(map[string]bool)
		for _, item := range items {
			if id := extractTimeEntryID(item.Text); id != "" {
				idSet[id] = true
			}
		}
		workItemsByIssue[issueID] = idSet
	}

	for _, entry := range timeEntries {
		displayDate := entry.startDate.Local().Format("Mon Jan 02 2006")

		if entry.issueID == "" {
			fmt.Printf("⚠️ No Issue: %s on %s\n", entry.description, displayDate)
			continue
		}

		existing := workItemsByIssue[entry.issueID]
		if existing[entry.id] {
			fmt.Printf("⏩ Skipping: %s (%s) on %s\n", entry.description, entry.issueID, displayDate)
			continue
		}

		args := postWorkItemArgs{
			description:     fmt.Sprintf("%s [%s]", entry.description, entry.id),
			dateMS:          entry.startDate.UnixMilli(),
			durationMinutes: entry.durationMinutes,
		}
		if entry.entryType != "" {
			if id, ok := workTypes[entry.entryType]; ok {
				args.workTypeID = id
			}
		}

		if !opts.dryRun {
			if err := youtrack.postIssueWorkItem(entry.issueID, args); err != nil {
				fmt.Fprintf(os.Stderr, "❌ Failed to sync %s (%s): %v\n", entry.description, entry.issueID, err)
				continue
			}
		}
		fmt.Printf("✅ Synced: %s (%s) on %s\n", entry.description, entry.issueID, displayDate)
	}

	return nil
}

func cleanUp(clockify *clockifyClient, youtrack *youtrackClient) error {
	projects, err := clockify.getWorkspaceProjects()
	if err != nil {
		return fmt.Errorf("fetch projects: %w", err)
	}

	for _, project := range projects {
		tasks, err := clockify.getActiveProjectTasks(project.ID)
		if err != nil {
			fmt.Fprintf(os.Stderr, "error fetching tasks for %s: %v\n", project.ID, err)
			continue
		}

		for _, task := range tasks {
			name, _ := task["name"].(string)
			issueID := extractIssueID(name)
			if issueID == "" {
				continue
			}

			issue, err := youtrack.getIssue(issueID)
			if err != nil {
				fmt.Fprintln(os.Stderr, err)
				continue
			}
			if issue.Resolved == nil {
				continue
			}

			body := make(map[string]any, len(task)+1)
			for k, v := range task {
				body[k] = v
			}
			body["status"] = "DONE"

			taskID, _ := task["id"].(string)
			if err := clockify.updateProjectTask(project.ID, taskID, body); err != nil {
				fmt.Fprintln(os.Stderr, err)
			}
		}
	}

	return nil
}
